File: AtCoder/ABC/182/cpp/E_Akari.cpp
#include <iostream>
#include <stdio.h>

#define MAX_H 1500
#define MAX_W 1500

using namespace std;
typedef long long ll;

int H, W, N, M;
int field[MAX_H][MAX_W];

int main() {
  cin >> H >> W >> N >> M;
  for (int i = 0; i < N; ++i) {
    int a, b;
    cin >> a >> b;
    field[a - 1][b - 1] = 1;
  }
  for (int i = 0; i < M; ++i) {
    int c, d;
    cin >> c >> d;
    field[c - 1][d - 1] = -1;
  }

  ll res = 0;
  for (int i = 0; i < H; ++i) {
    int j = 0;
    while (j < W) {
      while (j < W && field[i][j] != 1) {
        ++j;
      }
      while (j < W && field[i][j] == 1) {
        ++j;
      }
      while (j < W && (field[i][j] == 0 || field[i][j] == 2)) {
        if (field[i][j] == 0) {
          ++res;
          field[i][j] = 2;
        }
        ++j;
      }
    }
    --j;
    while (j >= 0) {
      while (j >= 0 && field[i][j] != 1) {
        --j;
      }
      while (j >= 0 && field[i][j] == 1) {
        --j;
      }
      while (j >= 0 && (field[i][j] == 0 || field[i][j] == 2)) {
        if (field[i][j] == 0) {
          ++res;
          field[i][j] = 2;
        }
        --j;
      }
    }
  }
  for (int j = 0; j < W; ++j) {
    int i = 0;
    while (i < H) {
      while (i < H && field[i][j] != 1) {
        ++i;
      }
      while (i < H && field[i][j] == 1) {
        ++i;
      }
      while (i < H && (field[i][j] == 0 || field[i][j] == 2)) {
        if (field[i][j] == 0) {
          ++res;
          field[i][j] = 2;
        }
        ++i;
      }
    }
    --i;
    while (i >= 0) {
      while (i >= 0 && field[i][j] != 1) {
        --i;
      }
      while (i >= 0 && field[i][j] == 1) {
        --i;
      }
      while (i >= 0 && (field[i][j] == 0 || field[i][j] == 2)) {
        if (field[i][j] == 0) {
          ++res;
          field[i][j] = 2;
        }
        --i;
      }
    }
  }
  printf("%lld\n", res + N);
}

File: AtCoder/ABC/182/python/E_Akari.py
H, W, N, M = map(int, input().split())

field = [[0] * W for _ in range(H)]
for _ in range(N):
    a, b = map(int, input().split())
    field[a - 1][b - 1] = 1

for _ in range(M):
    c, d = map(int, input().split())
    field[c - 1][d - 1] = -1


res = 0
for i in range(H):
    j = 0
    while j < W:
        while j < W and field[i][j] != 1:
            j += 1
        while j < W and field[i][j] == 1:
            j += 1
        while j < W and (field[i][j] == 0 or field[i][j] == 2):
            if field[i][j] == 0:
                res += 1
                field[i][j] = 2
            j += 1
    j = W - 1
    while j >= 0:
        while j >= 0 and field[i][j] != 1:
            j -= 1
        while j >= 0 and field[i][j] == 1:
            j -= 1
        while j >= 0 and (field[i][j] == 0 or field[i][j] == 2):
            if field[i][j] == 0:
                res += 1
                field[i][j] = 2
            j -= 1

for j in range(W):
    i = 0
    while i < H:
        while i < H and field[i][j] != 1:
            i += 1
        while i < H and field[i][j] == 1:
            i += 1
        while i < H and (field[i][j] == 0 or field[i][j] == 2):
            if field[i][j] == 0:
                res += 1
                field[i][j] = 2
            i += 1
    i = H - 1
    while i >= 0:
        while i >= 0 and field[i][j] != 1:
            i -= 1
        while i >= 0 and field[i][j] == 1:
            i -= 1
        while i >= 0 and (field[i][j] == 0 or field[i][j] == 2):
            if field[i][j] == 0:
                res += 1
                field[i][j] = 2
            i -= 1
print(res + N)
